File: backend/app/integrations/email.py
"""SMTP email delivery with a safe development logging fallback."""

from __future__ import annotations

from email.message import EmailMessage

import aiosmtplib

from app.core.config import Settings
from app.core.logging import get_logger

logger = get_logger(__name__)


class EmailService:
    """Send transactional authentication email without blocking the event loop."""

    def __init__(self, settings: Settings) -> None:
        self.settings = settings

    async def send(self, *, recipient: str, subject: str, body: str) -> None:
        """Deliver email through SMTP or log a development-only preview."""
        if not self.settings.smtp_host:
            logger.warning(
                "email_delivery_skipped",
                recipient=recipient,
                subject=subject,
                preview=body if self.settings.app_env == "development" else None,
            )
            return
        message = EmailMessage()
        message["From"] = self.settings.smtp_from_email
        message["To"] = recipient
        message["Subject"] = subject
        message.set_content(body)
        await aiosmtplib.send(
            message,
            hostname=self.settings.smtp_host,
            port=self.settings.smtp_port,
            username=self.settings.smtp_username,
            password=(
                self.settings.smtp_password.get_secret_value()
                if self.settings.smtp_password
                else None
            ),
            start_tls=self.settings.smtp_use_tls,
        )
        logger.info("email_sent", recipient=recipient, subject=subject)

File: backend/src/integrations/email.js
import nodemailer from "nodemailer";
import { config } from "../config.js";
import { log, logProcessFinished, logProcessStarted } from "../logger.js";

export class EmailService {
  constructor() {
    this.transport = config.smtpHost
      ? nodemailer.createTransport({
          host: config.smtpHost,
          port: config.smtpPort,
          secure: config.smtpPort === 465,
          requireTLS: config.smtpUseTls && config.smtpPort !== 465,
          auth: config.smtpUsername
            ? { user: config.smtpUsername, pass: config.smtpPassword }
            : undefined,
        })
      : null;
  }

  async send({ recipient, subject, body }) {
    logProcessStarted("Send email", { recipient, subject });
    if (!this.transport) {
      log("warn", "email_delivery_skipped", {
        recipient,
        subject,
        preview: config.appEnv === "development" ? body : undefined,
      });
      logProcessFinished("Send email", {
        recipient,
        subject,
        skipped: true,
      });
      return;
    }
    await this.transport.sendMail({
      from: config.smtpFromEmail,
      to: recipient,
      subject,
      text: body,
    });
    logProcessFinished("Send email", { recipient, subject, skipped: false });
  }
}

export const emailService = new EmailService();
